Exception Handling (#6)
Ada-Developers-Academy/core-unit-1: exception-handling/raising-and-handling-exceptions.md

# Exception Handling

> ## Learning Goals

Reviewer: *Reminder* for me tie in the sections of this lesson as right now it seems to jump around a bit. Things that would be great to add: 

- Providing real world examples of handling exceptions
- what it means to crash a program and why we want to prevent that,
- add questions.

- Define exceptions
- Explain how to handle exceptions using try and except clauses

## Introduction

If you didn't know, the absolute worst-case scenario for software is a crash caused by a syntax or runtime error*. The software has gone to a state where the program can't do anything, the user can't do anything, and possibly the program can't even restart into a usable state!

> If software fails for one person, it's likely that software is failing for _many_ people. All of these consequences result in 

Author: LOLLLLLL I JUST SAW THIS. Why did I not finish that sentence? Who knows, but THIS SO SO MUCH BETTER, THANK YOU! 😆

```

Try Clause
If any exception is raised inside the `try` clause, the rest of the `try` clause is skipped, and code execution moves to the `except` clause
Except Clause, with Exception Type next to it
A raised exception from the `try` clause will check if it matches the exception type next to the `except` keyword. If it matches, the `except` clause executes
`as var_name:`
Inside the `except` clause, it might be helpful to refer to the Exception itself. For example, we might want to `print` it. We use `as var_name` to say the exception is in the variable `var_name`. `var_name` can be any name.

Example:

```python

```

The code execution of the code above is as follows:

1. First, the Python interpreter says "Okay, the function ` ` is defined."
1. Then, the Python interpreter calls ` `
1. We enter the `try` clause
1. Then, the Python interpreter executes ` `. This line of code raises an exception
1. We skip the rest of the `try` clause, and then check if the exception is a ` `
1. The exception is a ` `, so we enter the `except` clause. There is a new variable ` `, whose value is the Exception raised.
1. Then, we print the exception with a message. We could do other stuff too!

> 1. Finally, we exit this whole thing.

Author: You are the best.

> 1. Finally, we exit this whole thing.

Author: Just the literal best.

More Examples:

```python

```

```python

```

### Handle Many Types of Exceptions

If we need to handle more than one kind of exception, we can add an infinite number of `except` clauses. Much like the `if..elif` statements, the raised exception will check if it matches one-at-a-time, starting from the top.

This example will run the `try` clause. If an exception is raised, it will check if the exception is a ` ` first. If it isn't, then it will check if it's ` `. Finally, if it isn't ` `, then it will check if it's ` `. 

```python

```

## We Can Define Exceptions

If none of the built-in functions don't suit the logic and context of our code, what can we do? We can define custom exceptions, or exceptions with our own names and implementations.

We won't cover that material, but it's a cool piece of syntax! Follow your curiosity!

## Summary

> 

Reviewer: ```suggestion
Whether they're caused by programmers, insufficient memory, or user inputs errors are bound to happen. Raising and using the `try..except` clause are great ways to make our code more robust by handling errors before they crash our program. 
```

## Check for Understanding

> 

Reviewer 2: <!-- BEGIN CHALLENGE -->

### !challenge

* type: ordering
* id: e45ab433-76f7-46b8-8203-92f7df1a7515
* title: try-except block

##### !question

Assume we have a function process_list which takes a list as an argument and performs an operation on it.  If the function is passed something other than a list, it throws a TypeError exception.  It may throw another type of exception if something else goes wrong.  Re-order the following lines of code into a working try-except block.  Please ignore the lack of proper indentation.

##### !end-question

##### !answer

1. try:
1.      process_list(my_list)
1. except TypeError:
1.      print("Variable is not a list")
1. except:
1.      print("Something else went wrong")

##### !end-answer

### !end-challenge

<!-- END CHALLENGE -->

Author: I needed to manually add this in because GH suggestions don't play well with these.

Tagging @jbieniosek for visibility! No action needed, just wanted to make sure you knew I took this suggestion.

## Exceptions Are Python's Cry For Help

An exception is a type of _thing_ (we can say "data type") that holds information about an error that the program encounters. Exceptions hold data like specifically what error occured and which line of code caused it.

Python creates an exception whenever a syntax or runtime error occurs in a program. Python will then immediately try to stop the execution of the program and exit out of it. When an error happens, and we see details in the stack trace, we actually see details about the exception!

The act of creating an exception is often called **raising an exception.**

When an exception is raised, Python moves "up" through the stack of code execution. This act of the exception moving "up" is often called **throwing an exception** or **bubbling up.** We might imagine an exception raised in a deep layer of code gets thrown up through the each layer of code, eventually to the programmer's face.

Python raises exceptions because exceptions are Python's cry for help. Python only raises exceptions when some code says, "I can't do this! I can't go on!"

> Of course, when Python surrenders and throws an exception, we can't force it to keep going. Ideally, however, should we see an exception, we'd love to recognize it, handle the situation, and then allow our program to move on.

Author: Awesome, I appreciate the logical catch!!

> Code that recognizes a thrown exception is often called **catching an exception.** Catching an exception and then handling the situation, setting things back in order, and making it possible to move on, is often called **handling an exception** or **rescuing an exception.**

Reviewer 3: Suggest avoiding the ruby terminology.
```suggestion
Code that recognizes a thrown exception is often called **catching an exception.** Catching an exception and then handling the situation, setting things back in order, and making it possible to move on, is also often called **handling an exception.**
```

Author: Didn't realize that was only a rubyism, but now that I think about it, I didn't learn "rescue" until Ruby haha. thanks for the catch!

> It is wise and ideal to write code that rescues and helps out our program, should Python ever raise an exception.

Author: perfect awesome lead-in to learning!

## Common Exception Types

| Exception Name    | When it occurs                                                                      |
| ----------------- | ----------------------------------------------------------------------------------- |
| NameError         | Raised when a local or global name is not found.                                    |
| ZeroDivisionError | Raised when the second argument of a division or modulo operation is zero.          |
| OverflowError     | Raised when the result of an arithmetic operation is too large to be represented.   |
| SyntaxError       | Raised when the parser encounters a syntax error.                                   |
| TypeError         | Raised when an operation or function is applied to an object of inappropriate type. |

> | KeyboardInterrupt | Raised when the user hits the interrupt key (normally Control-C or Delete).         |

Author: I'll add/adjust this manually; I think since the point of the passage is to communicate "these exceptions exist" (and handling has barely been talked about), I'll go for simpler examples

> | KeyboardInterrupt | Raised when the user hits the interrupt key (normally Control-C or Delete).         |

Reviewer 3: Yeah, there's definitely a risk of front loading so much in an overview that the reader shuts down.

## We Can Raise Exceptions

Python can raise exceptions under the hood... and other packages and tests can raise exceptions...

> We can too! We use the `raise` keyword. We pair the `raise` keyword with the exception class (type) that we are expecting.

Author: Nice, i like this

```python
x = -1
if x < 1:
    raise ValueError
```

When we run this code, we see the following output in the console:

```
Traceback (most recent call last):
  File "main.py", line 3, in <module>
    raise ValueError
ValueError
```

Here, we should observe what our stack trace is telling us:

- There is a `ValueError` that stopped the code execution
- The line of code that caused this error is `raise ValueError`
- This line of code is in `"main.py"` on line 3

### Creating Error Messages

We can actually pass in our own error message when we raise an exception! If we pass in an error message, the exception will use that string as its error description when needed.

```python
raise ZeroDivisionError('Tried to divide by zero flowers.')
```

When we run this code, we see the following output, which now has our message about flowers!

```
Traceback (most recent call last):
  File "main.py", line 1, in <module>
    raise ZeroDivisionError('Tried to divide by zero flowers.')
ZeroDivisionError: Tried to divide by zero flowers.
```

## We Can Rescue Any Raised Exception

Remember, the worst case scenario for any program and programmer is a crash, especially from a raised exception.

> What can we do to protect our program from crashes? Of course, the first step is to review the logic, and see if there's a way to write it that won't raise errors. However, Python also lets us catch raised exceptions before it crashes our program-- and then handle them! We use the `try ... except` syntax for this.

Author: i wish i could give you an award for this.

this feels like waaaaay better and more relevant wisdom, thank you super much!

| `error_as_a_variable`:    | **Replace this** with any valid variable name. This variable's value will be the exception data, and can be accessed inside of the except-clause. `err` is a common variable name. Don't forget the `:`!           |
| Body of the except clause | The error clause should include code that needs to execute if an exception is raised. This code usually is used for printing details about `err`, and/or "rescuing" the situation and helping the program move on. |

### !callout-info

## Specifying Error Type is Optional...

Technically, specifying the error type is optional. Specifying the local variable `err` is also optional. From the above syntax, we could omit `ExampleError as error_as_a_variable`, and it would still be valid code. Then, that except-clause would run if it catches _any_ exception. It's valuable to have the specificity and detail when handling exceptions, so this is a pattern we rarely see.

### !end-callout

### Walking Through an Example: Circumference

Consider this example. First, read through the code and its console output. Then, read through the explanation, and use it to trace the code.

```python

> def calculate_circumference(radius):

Author: I'll check!

There's a `hasn't` in a string below; leaving this as-is for now.

    except NameError as err:
        print(f"We're trying to print carrot, but carrot is never defined before this. {err}")

    print("**************")

# The following line raises a ZeroDivisionError
do_weird_number_math(3)

# The following line raises an UnboundLocalError
do_weird_number_math(5555)

# This line raises a NameError
do_weird_number_math(8)
```

This code produces this console output:

```
We're entering the do_weird_number_math function! Value of apple: 3
apple is 3, so it tried to divide by zero. division by zero
**************
We're entering the do_weird_number_math function! Value of apple: 5555
apple is valid, but banana was never given a value, so we get an error. local variable 'banana' referenced before assignment
**************
We're entering the do_weird_number_math function! Value of apple: 8
Value of banana: 1.6
We're trying to print carrot, but carrot is never defined before this. name 'carrot' is not defined
**************
```

In our try-clause, different situations will raise different exceptions. Chaining except-clauses allows us to rescue them in specific ways.

> 

Author: In the interest of space/priority, gonna turn this into a callout and go less in-depth

> 

Reviewer 3: Sounds good. Glad we can squeeze something about this in here. I know students are getting a ton of information, but it seemed relevant enough to the material here.
